File: solutions/1382/1382-kir3i.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root):
        vals = []
        self.popNode(root, vals)
        return self.makeTree(0, len(vals)-1, vals)

    def makeTree(self, lo, hi, vals):
        if lo > hi:
            return None
        elif lo == hi:
            return TreeNode(vals[lo])
        mid = (lo + hi) // 2
        root = TreeNode(vals[mid])
        root.left = self.makeTree(lo, mid-1, vals)
        root.right = self.makeTree(mid+1, hi, vals)

        return root

    def popNode(self, root, res):
        if root.left:
            self.popNode(root.left, res)
        res.append(root.val)
        if root.right:
            self.popNode(root.right, res)

File: solutions/1382/1382-kir3i.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> vals;
    TreeNode* balanceBST(TreeNode* root) {
        popNode(root);
        return makeTree(0, vals.size() - 1);
    }
    
    TreeNode* makeTree(int lo, int hi) {
        if (lo > hi)        return nullptr;
        else if (lo == hi)  return new TreeNode(vals[lo]);
        
        int mid = (lo + hi) / 2;
        TreeNode* root = new TreeNode(vals[mid]);
        root->left = makeTree(lo, mid - 1);
        root->right = makeTree(mid + 1, hi);
        return root;
    }
    
    void popNode(TreeNode *root) {
        if(!root)   return;
        popNode(root->left);
        vals.push_back(root->val);
        popNode(root->right);
    }
};
